### DISCUSSION TITLE
createData in OAuth2WithCode

### DISCUSSION THREAD

**Discussion by gaganbiswas** [pocketbase/pocketbase#4801] [Q&A] [answered]
```js
const authData = await pb.client
      .collection("users")
      .authWithOAuth2Code(
        name,
        paramCode!,
        codeVerifier,
        process.env.CALLBACK_URL!,
        {
          emailVisibility: true,
        }
      );

    if (authData?.token) {
      const { record, token } = authData;
      record.token = token;
      cookies().set("pb_auth", pb.client.authStore.exportToCookie(), {
        httpOnly: true,
      });

      if (authData.meta?.isNew) {
        const name =
          authData.meta?.name ||
          authData.meta?.rawUser?.name ||
          authData.meta?.rawUser?.given_name +
            " " +
            authData.meta?.rawUser?.family_name;

        const avatar =
          authData.meta?.avatarUrl || authData.meta?.rawUser?.picture;

        await pb.client.collection("users").update(
          authData.record.id,
          {
            name: name.trim(),
            avatar_url: avatar,
          },
          {
            headers: {
              Authorization: process.env.AUTH_TOKEN!,
            },
          }
        );
      }
```

On the first login using google I am not getting the name and avatar in my cookie. since the cookie is getting set from the token returned by the authWithOAuth which at that point doesn't have the values.

Is there a way to put the name and the avatar in the cookie using custom createData in `authWithOAuth2Code()` which currently just has emailVisibility set to true only.

**Comment by ganigeorgiev** [comment, marked as answer]
> Is there a way to put the name and the avatar in the cookie using custom createData

No. Move your `cookies().set(...)` call after the update.

_Side-note:_ the `cookie.set` call is not "correct". I've seen this error several times and I'm not sure from where everyone seems to copy it. JS SSR comes with a lot of complications when integrating with external backends like PocketBase and I don't recommend it if you are new to web development and not sure whether you need it (_you can read some ramblings on this in https://github.com/pocketbase/pocketbase/discussions/3792#discussioncomment-7620032_). PocketBase is designed to be interact directly and that is the easiest and most performant way to use it.
The SDK `loadFromCookie` and `exportToCookie` methods operates with raw cookies, aka. `exportToCookie` returns a plain cookie string (eg. `pb_auth=...`, including the appropiate attributes like sameSite, maxAge, etc.).
I assume that in the above `cookies()` is the SvelteKit or some other framework cookies abstraction. In that case, if you are going to use it then you don't need `exportFromCookie` because in your code you are practically setting a cookie string as value to another cookie that doesn't seem to even have an expiration. Your cookie value could be simply `JSON.stringify({token: pb.authStore.token})`.
**Keep in mind that neither the JS SDK, nor the PocketBase server, use cookies.** The cookie is there just to share the auth state between the client/browser and your node server.

**Reply by gaganbiswas** [reply]
Thanks for the information and the recommendation and the detailed insight on this. I will look into the cookie part right away. I never actually looked into this since I am using this in a side project.

**Reply by gaganbiswas** [reply]
@ganigeorgiev 

```js
const cookieJSON = cookieStringToJSON(pb.client.authStore.exportToCookie());

cookies().set("pb_auth", cookieJSON["pb_auth"], {
     httpOnly: true,
     secure: process.env.NODE_ENV === "production" ? true : false,
     expires: new Date(cookieJSON["Expires"]),
});
```

Did something like this. The function `cookieStringToJSON()` takes in a cookie string and returns the json.

```js
const cookieStringToJSON = (cookieString: string) => {
  const components = cookieString.split("; ");
  const cookieObj: any = {};

  components.forEach(function (part: string) {
    let [key, value] = part.split("=");
    cookieObj[key.trim()] = value ? value.trim() : true;
  });

  return cookieObj;
};
```

I hope this makes it a bit better than what it was. Given now it has an expiration date and i am storing the value of pb_auth inside a cookie called pb_auth, instead of storing the whole cookie string as value to another cookie.

_Side note: I am actually working with Next.js SSR and not svelte. When getting the data from the database I am actually using the Pocketbase SDK in the server side to do it. Only the modified part is I guess this authentication stuff._

**Reply by ganigeorgiev** [reply]
I guess it is ok but I think it is unnecessary as you don't need to parse the `exportToCookie` string.

If you are using an external cookies abstraction, then you are in full control of its value. In your case it could be simply `JSON.stringify({ token: pb.authStore.token })` (_eventually you can also export the model if needed_).

For the reverse operation, to populate the SDK store from a cookie, you can either use `loadFromCookie` if you have access to the raw cookie string, or just call `pb.authStore.save(token, model)` with the extracted token and model value from the cookie using [`cookies.get()`](https://nextjs.org/docs/app/api-reference/functions/cookies#cookiesgetname).

**Reply by ganigeorgiev** [reply]
I forgot to mention that you can also use the SDK `cookieParse` helper if you really need it:

```js
import PocketBase, { cookieParse } from "pocketbase"

const result = cookieParse(cookieStr)
```

**Reply by gaganbiswas** [reply]
> I forgot to mention that you can also use the SDK `cookieParse` helper if you really need it:
> 
> 
> 
> ```js
> 
> import PocketBase, { cookieParse } from "pocketbase"
> 
> 
> 
> const result = cookieParse(cookieStr)
> 
> ```

Thanks for this again! I will try both of the above tomorrow. The only reason I am parsing the cookie string is to get the expire time for the token(i guess it is the token expiry time). In that case, the cookie that i am creating will have the same expiry time as the token.

I guess the 'authStore.save(token, model)' would work great, but in that case I have to set a custom expiry time for my token(and model) cookies.
